.doom.d/doom_config.org
* Init
:PROPERTIES:
:header-args: :tangle ~/.doom.d/init.el
:END:

#+begin_src elisp
;;; init.el -*- lexical-binding: t; -*-

;; Remember to run 'doom refresh' after modifying it.
;;
;; More information about these modules (and what flags they support) can be
;; found in modules/README.org.

(doom! :input
       ;;chinese
       ;;japanese

       :completion
       company           ; the ultimate code completion backend
                                        ;(helm
                                        ; +fuzzy
                                        ; :lang org)              ; the *other* search engine for love and life
       ;;ido               ; the other *other* search engine...
       (ivy +fuzzy)               ; a search engine for love and life

       :ui
       ;;deft              ; notational velocity for Emacs
       doom              ; what makes DOOM look the way it does
       doom-dashboard    ; a nifty splash screen for Emacs
       (emoji
        +github
        +unicode)
       ;; doom-quit         ; DOOM quit-message prompts when you quit Emacs
       ;; fill-column       ; a `fill-column' indicator
       hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
       hydra
       indent-guides     ; highlighted indent columns
       modeline          ; snazzy, Atom-inspired modeline, plus API
       nav-flash         ; blink the current line after jumping
       ;;neotree           ; a project drawer, like NERDTree for vim
       ophints           ; highlight the region an operation acts on
       (popup            ; tame sudden yet inevitable temporary windows
        +all             ; catch all popups that start with an asterix
        +defaults)       ; default popup rules
       ;; (pretty-code
       ;; +ioveska)       ; replace bits of code with pretty symbols
       ;;tabs              ; an tab bar for Emacs
       treemacs          ; a project drawer, like neotree but cooler
       unicode           ; extended unicode support for various languages
       vc-gutter         ; vcs diff in the fringe
       ;; vi-tilde-fringe   ; fringe tildes to mark beyond EOB
       ;; window-select     ; visually switch windows
       workspaces        ; tab emulation, persistence & separate workspaces
       zen

       :editor
       ;;(evil +everywhere); come to the dark side, we have cookies
       ;; file-templates    ; auto-snippets for empty files
       ;;god               ; run Emacs commands without modifier keys
       fold              ; (nigh) universal code folding
       (format +onsave)  ; automated prettiness
       ;;lispy             ; vim for lisp, for people who dont like vim
       ;;multiple-cursors  ; editing in many places at once
       ;;objed             ; text object editing for the innocent
       ;;parinfer          ; turn lisp into python, sort of
       ;; rotate-text       ; cycle region at point between text candidates
       snippets          ; my elves. They type so I don't have to
       word-wrap         ; soft wrapping with language-aware indent

       :emacs
       (dired +icons)    ; making dired pretty [functional]
       electric          ; smarter, keyword-based electric-indent
       vc                ; version-control and Emacs, sitting in a tree
       undo
       ibuffer

       :term
       ;;eshell            ; a consistent, cross-platform shell (WIP)
       ;;shell             ; a terminal REPL for Emacs
       term              ; terminals in Emacs
       ;;vterm             ; another terminals in Emacs

       :checkers
       ;; grammar
       spell
       syntax

       :tools
       ;;ansible
       debugger          ; FIXME stepping through code, to help you add bugs
       ;;direnv
       docker
       editorconfig      ; let someone else argue about tabs vs spaces
       ;; ein               ; tame Jupyter notebooks with emacs
       eval              ; run code, run (also, repls)
       ;;gist              ; interacting with github gists
       (lookup           ; helps you navigate your code and documentation
        +docsets
        +dictionary
        +offline)        ; ...or in Dash docsets locally
       ;; lsp
       (magit +forge)             ; a git porcelain for Emacs
       make              ; run make tasks from Emacs
       ;;pass              ; password manager for nerds
       pdf               ; pdf enhancements
       ;;prodigy           ; FIXME managing external services & code builders
       ;;rgb               ; creating color strings
       terraform         ; infrastructure as code
       ;;tmux              ; an API for interacting with tmux
       ;;upload            ; map local to remote projects via ssh/ftp
       ;;wakatime

       :lang
       ;;agda              ; types of types of types of types...
       ;;assembly          ; assembly for fun or debugging
       ;;(cc)                ; C/C++/Obj-C madness
       ;;clojure           ; java with a lisp
       ;;common-lisp       ; if you've seen one lisp, you've seen them all
       ;;coq               ; proofs-as-programs
       ;;crystal           ; ruby at the speed of c
       ;;csharp            ; unity, .NET, and mono shenanigans
       data              ; config/data formats
       ;;erlang            ; an elegant language for a more civilized age
       ;;elixir            ; erlang done right
       ;;elm               ; care for a cup of TEA?
       emacs-lisp        ; drown in parentheses
       ;;       ess               ; emacs speaks statistics
       ;;fsharp           ; ML stands for Microsoft's Language
       ;;go                ; the hipster dialect
       ;;(haskell +intero) ; a language that's lazier than I am
       ;;hy                ; readability of scheme w/ speed of python
       ;;idris             ;
       ;;(java +meghanada) ; the poster child for carpal tunnel syndrome
       ;;javascript        ; all(hope(abandon(ye(who(enter(here))))))
       ;;julia             ; a better, faster MATLAB
       ;;kotlin            ; a better, slicker Java(Script)
       ;;latex             ; writing papers in Emacs has never been so fun
       ;;lean
       ;;ledger            ; an accounting system in Emacs
       ;;lua               ; one-based indices? one-based indices
       markdown          ; writing docs for people to ignore
       ;;nim               ; python + lisp at the speed of c
       ;;nix               ; I hereby declare "nix geht mehr!"
       ;;ocaml             ; an objective camel
       (org              ; organize your plain life in plain text
        ;;+dragndrop       ; drag & drop files/images into org buffers
        +pandoc          ; export-with-pandoc support
        +jupyter
        +roam
        +pretty
        ;;+present
        )        ; using org-mode for presentations
       ;;perl              ; write code no one else can comprehend
       ;;php               ; perl's insecure younger brother
       ;;plantuml          ; diagrams for confusing people more
       ;;purescript        ; javascript, but functional
       (python
        +pyenv
        +poetry)            ; beautiful is better than ugly
       ;;qt                ; the 'cutest' gui framework ever
       racket            ; a DSL for DSLs
       ;; rest              ; Emacs as a REST client
       ;;ruby              ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
       ;;rust              ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
       ;;scala             ; java, but good
       ;;scheme            ; a fully conniving family of lisps
       sh                ; she sells {ba,z,fi}sh shells on the C xor
       ;;solidity          ; do you need a blockchain? No.
       ;;swift             ; who asked for emoji variables?
       ;;terra             ; Earth and Moon in alignment for performance.
       ;;web               ; the tubes
       yaml

       :email
       ;;(mu4e +gmail)       ; WIP
       ;;notmuch             ; WIP
       ;;(wanderlust +gmail) ; WIP

       ;; Applications are complex and opinionated modules that transform Emacs
       ;; toward a specific purpose. They may have additional dependencies and
       ;; should be loaded late.
       :app
       ;;calendar
       ;;irc               ; how neckbeards socialize
       (rss +org)        ; emacs as an RSS reader
       ;;twitter           ; twitter client https://twitter.com/vnought

       :os
       macos             ; MacOS-specific commands

       :config
       (default +bindings)
       )
#+end_src

* Config
:PROPERTIES:
:header-args: :tangle ~/.doom.d/config.el
:END:
#+begin_src elisp
;;; .doom.d/config.el -*- lexical-binding: t; -*-
#+end_src
** Coding
*** Code folding
#+begin_src elisp
(define-key prog-mode-map (kbd "C-<tab>") '+fold/toggle)
(define-key prog-mode-map (kbd "C-<") '+fold/close-all)
(define-key prog-mode-map (kbd "C->") '+fold/open-all)
#+end_src

*** Python / Jupyter
#+begin_src elisp
(after! python
  (setq-default flycheck-disabled-checkers '(python-pylint))
  (add-hook! 'python-mode-hook
             :append (anaconda-mode)
             :append (python-docstring-mode)
             )
  )

(after! ob-jupyter
  (org-babel-jupyter-override-src-block "python")
  )
#+end_src

#+RESULTS:

*** Magit
#+begin_src elisp
(after! magit
  (setq magit-commit-show-diff nil)
  )
#+end_src


** Moving / Loading
*** Ivy
#+begin_src elisp
(after! ivy
  (define-key ivy-minibuffer-map (kbd "<left>") 'counsel-up-directory)
  (define-key ivy-minibuffer-map (kbd "<right>") 'ivy-alt-done)
  (define-key ivy-minibuffer-map (kbd "C-<return>") 'ivy-immediate-done)
  )
#+end_src

*** Framemove
Move easily across frames with s-<arrow-key>.

#+name: framemove
#+begin_src elisp
(push (expand-file-name "~/repos/dotfiles/elisp") load-path)
(require 'framemove)
(windmove-default-keybindings 'super)
(setq framemove-hook-into-windmove t)
#+end_src

*** Avy
Make avy work across all windows and add quick bindings for avy goto.
#+name: avy
#+begin_src elisp
(setq avy-all-windows 'all-frames)
(map! :leader
      (:prefix ("m" . "move")
       :desc "by word/subword" "w" #'avy-goto-word-or-subword-1
       :desc "by char" "c" #'avy-goto-char
       :desc "by many characters" "t" #'avy-goto-char-timer
       )
      )
#+end_src


** Writing / Reading / Drawing

*** Org
**** General
This does a few things:
- change the agenda files to point to one todo file
- create a default file for notes
- update todo keywords
- enable auto fill
- display inline images
- indent org files automatically
#+begin_src elisp
(after! org
  (setq
   org-confirm-babel-evaluate nil
   org-agenda-files '("/Users/virgile/org/todo.org")
   org-default-notes-file (expand-file-name "~/org/notes.org")
   org-todo-keywords '((sequence "TODO" "DOING" "|" "DONE" "ARCHIVED")
                       (sequence "QUESTION" "WORKING-ON-IT" "|" "ANSWERED"))
   )
  (add-hook! 'org-mode-hook :append #'turn-on-auto-fill #'org-indent-mode)
  (add-hook 'org-babel-after-execute-hook 'org-display-inline-images)
  )
#+end_src

**** Roam
#+begin_src elisp
(after! org
  (setq
   org-roam-graph-viewer "open"
   org-roam-dailies-directory "daily/"
   org-roam-dailies-capture-templates '(("d" "default" entry
                                         #'org-roam-capture--get-point
                                         "* %?"
                                         :file-name "daily/%<%Y-%m-%d>"
                                         :head "#+TITLE: %<%Y-%m-%d>\n\n* Check-in [/]\n"))
   org-roam-graph-exclude-matcher '("private" "daily" "jira-tickets")
   )
  )
#+end_src

#+RESULTS:
| private | daily | jira-tickets |

**** Bindings
#+begin_src elisp
(after! org
  (map! :map org-mode-map
        "C-<return>" #'org-insert-heading
        "C-M-<right>" #'org-demote-subtree
        "C-M-<left>" #'org-promote-subtree
        )

  (map! :map org-mode-map :leader
        (:prefix-map ("i" . "insert")
         :desc "drawer" "d" #'org-insert-drawer
         :desc "heading" "h" #'org-insert-heading
         :desc "item" "i" #'org-insert-item
         :desc "link" "l" #'org-insert-link
         :desc "subheading" "s" #'org-insert-subheading
         :desc "template" "t" #'org-insert-structure-template
         ))
  )
#+end_src

**** Additional export mode
#+begin_src elisp
(after! org
  (require 'ox-gfm nil t)
  (require 'ox-twbs nil t)
  (require 'ox-rst nil t)
  )
#+end_src

*** Elfeed
#+begin_src elisp
(after! elfeed
  (setq-default elfeed-search-filter "@2-days-ago -work")
  (map! :map elfeed-show-mode-map
        "a" #'pocket-reader-add-link)
  (map! :map elfeed-search-mode-map
        "a" #'pocket-reader-add-link)
  )
#+end_src

*** Zen mode
Turn off text zooming
#+begin_src elisp
(setq +zen-text-scale 0)
#+end_src

#+RESULTS:
: 0

Increase width
#+begin_src elisp
(setq writeroom-width 120)
#+end_src

*** Mermaid
#+begin_src elisp
(setq ob-mermaid-cli-path (executable-find "mmdc"))
#+end_src

*** Treemacs
#+begin_src elisp
(after! treemacs
  (map! :map treemacs-mode-map
        "SPC" #'treemacs-peek)
  )

(map! :leader
      :desc "treemacs" "t t" #'treemacs)
#+end_src

#+RESULTS:
: treemacs

** Global changes
*** Key bindings
#+begin_src elisp
(map! [home] #'move-beginning-of-line
      [end] #'move-end-of-line
      "C-x g" #'magit-status
      "s-k" #'doom/kill-this-buffer-in-all-windows
      "C-/" #'company-filter-candidates
      )
#+end_src

#+RESULTS:

Remove default smartparens bindings.
#+begin_src elisp
(after! smartparens
  (map! :map smartparens-mode-map
        "C-<left>" nil
        "C-<right>" nil
        "M-<left>" nil
        "M-<right>" nil)
  )
#+end_src
*** Hydras
**** Resize window
#+begin_src elisp
(defhydra hydra-window-resizing (:hint nil)
  "
Resizing frames
---------------
  [→] + horizontal
  [←] - horizontal
  [↑] + vertical
  [↓] - vertical
"
  ("<up>" enlarge-window)
  ("<down>" shrink-window)
  ("<right>" enlarge-window-horizontally)
  ("<left>" shrink-window-horizontally)
  )
#+end_src

**** Bindings
#+begin_src elisp
(map! :leader
      (:prefix-map ("h" . "hydras")
       :desc "buffer resizing" "r" #'hydra-window-resizing/body
       :desc "jupyter" "j" #'jupyter-org-hydra/body
       :desc "smerge" "m" #'+vc/smerge-hydra/body
       :desc "zoom" "z" #'+hydra/text-zoom/body
       ))

(map! :map smerge-mode-map
      "s-m" #'+vc/smerge-hydra/body)

(map! :map org-mode-map
      "s-h" #'jupyter-org-hydra/body)
#+end_src
*** Modes
Turn on wrap mode
#+begin_src elisp
(+global-word-wrap-mode 1)
#+end_src

Allow camel case navigation *everywhere*
#+begin_src elisp
(global-subword-mode)
#+end_src

*** Others
Display time in current frame
#+begin_src elisp
(display-time)
#+end_src

Display directory name in buffer name when file names are the same
#+begin_src elisp
(setq uniquify-buffer-name-style 'post-forward-angle-brackets)
#+end_src

Turn off confirmation message at exit time
#+begin_src elisp
(setq confirm-kill-emacs nil)
#+end_src

Increase garbage collector limit
#+begin_src elisp
(after! gcmh
  (setq gcmh-high-cons-threshold 33554432))
#+end_src

Doom popups
#+begin_src elisp
(map! :leader
      :desc "Show popup" "t p" #'+popup/toggle
      )
#+end_src

** Theme
*** Change theme
#+begin_src elisp
(load-theme 'doom-palenight t)
(setq doom-themes-enable-bold nil)
(set-face-bold-p 'bold nil)
#+end_src

#+RESULTS:

*** Customize fonts
#+begin_src elisp
(setq mixed-pitch-set-height t)
(setq doom-font (font-spec :family "Hack" :size 13)
      doom-variable-pitch-font (font-spec :family "Lora" :size 16)
      )
(doom/reload-font)
#+end_src

#+RESULTS:
| #<frame doom_config.org – Doom Emacs 0x7f95eb056520> |

* Packages
:PROPERTIES:
:header-args: :tangle ~/.doom.d/packages.el
:END:

#+begin_src elisp
(package! ox-gfm)
(package! ox-twbs)
(package! ox-rst)
(package! pocket-reader)
(package! zoom)
(package! ob-mermaid)
(package! mermaid-mode)
(package! python-docstring)
#+end_src

* Local Variables
# Local Variables:
# eval: (add-hook 'after-save-hook (lambda ()(org-babel-tangle)) nil t)
# End:


.doom.d/config.el
;;; .doom.d/config.el -*- lexical-binding: t; -*-

(define-key prog-mode-map (kbd "C-<tab>") '+fold/toggle)
(define-key prog-mode-map (kbd "C-<") '+fold/close-all)
(define-key prog-mode-map (kbd "C->") '+fold/open-all)

(after! python
  (setq-default flycheck-disabled-checkers '(python-pylint))
  (add-hook! 'python-mode-hook
             :append (anaconda-mode)
             :append (python-docstring-mode)
             )
  )

(after! ob-jupyter
  (org-babel-jupyter-override-src-block "python")
  )

(after! magit
  (setq magit-commit-show-diff nil)
  )

(after! ivy
  (define-key ivy-minibuffer-map (kbd "<left>") 'counsel-up-directory)
  (define-key ivy-minibuffer-map (kbd "<right>") 'ivy-alt-done)
  (define-key ivy-minibuffer-map (kbd "C-<return>") 'ivy-immediate-done)
  )

(push (expand-file-name "~/repos/dotfiles/elisp") load-path)
(require 'framemove)
(windmove-default-keybindings 'super)
(setq framemove-hook-into-windmove t)

(setq avy-all-windows 'all-frames)
(map! :leader
      (:prefix ("m" . "move")
       :desc "by word/subword" "w" #'avy-goto-word-or-subword-1
       :desc "by char" "c" #'avy-goto-char
       :desc "by many characters" "t" #'avy-goto-char-timer
       )
      )

(after! org
  (setq
   org-confirm-babel-evaluate nil
   org-agenda-files '("/Users/virgile/org/todo.org")
   org-default-notes-file (expand-file-name "~/org/notes.org")
   org-todo-keywords '((sequence "TODO" "DOING" "|" "DONE" "ARCHIVED")
                       (sequence "QUESTION" "WORKING-ON-IT" "|" "ANSWERED"))
   )
  (add-hook! 'org-mode-hook :append #'turn-on-auto-fill #'org-indent-mode)
  (add-hook 'org-babel-after-execute-hook 'org-display-inline-images)
  )

(after! org
  (setq
   org-roam-graph-viewer "open"
   org-roam-dailies-capture-templates '(("d"
                                         "daily"
                                         plain
                                         (function org-roam-capture--get-point)
                                         :immediate-finish t
                                         :file-name "dailies/%<%Y-%m-%d>"
                                         :head "#+TITLE: %<%Y-%m-%d>"))
   org-roam-graph-exclude-matcher '("private" "dailies" "jira-tickets")
   )
  )

(after! org
  (map! :map org-mode-map
        "C-<return>" #'org-insert-heading
        "C-M-<right>" #'org-demote-subtree
        "C-M-<left>" #'org-promote-subtree
        )

  (map! :map org-mode-map :leader
        (:prefix-map ("i" . "insert")
         :desc "drawer" "d" #'org-insert-drawer
         :desc "heading" "h" #'org-insert-heading
         :desc "item" "i" #'org-insert-item
         :desc "link" "l" #'org-insert-link
         :desc "subheading" "s" #'org-insert-subheading
         :desc "template" "t" #'org-insert-structure-template
         ))
  )

(after! org
  (require 'ox-gfm nil t)
  (require 'ox-twbs nil t)
  (require 'ox-rst nil t)
  )

(after! elfeed
  (setq-default elfeed-search-filter "@2-days-ago -work")
  (map! :map elfeed-show-mode-map
        "a" #'pocket-reader-add-link)
  (map! :map elfeed-search-mode-map
        "a" #'pocket-reader-add-link)
  )

(setq +zen-text-scale 0)

(setq writeroom-width 120)

(setq ob-mermaid-cli-path (executable-find "mmdc"))

(after! treemacs
  (map! :map treemacs-mode-map
        "SPC" #'treemacs-peek)
  )

(map! :leader
      :desc "treemacs" "t t" #'treemacs)

(map! [home] #'move-beginning-of-line
      [end] #'move-end-of-line
      "C-x g" #'magit-status
      "s-k" #'doom/kill-this-buffer-in-all-windows
      "C-/" #'company-filter-candidates
      )

(after! smartparens
  (map! :map smartparens-mode-map
        "C-<left>" nil
        "C-<right>" nil
        "M-<left>" nil
        "M-<right>" nil)
  )

(defhydra hydra-window-resizing (:hint nil)
  "
Resizing frames
---------------
  [→] + horizontal
  [←] - horizontal
  [↑] + vertical
  [↓] - vertical
"
  ("<up>" enlarge-window)
  ("<down>" shrink-window)
  ("<right>" enlarge-window-horizontally)
  ("<left>" shrink-window-horizontally)
  )

(map! :leader
      (:prefix-map ("h" . "hydras")
       :desc "buffer resizing" "r" #'hydra-window-resizing/body
       :desc "jupyter" "j" #'jupyter-org-hydra/body
       :desc "smerge" "m" #'+vc/smerge-hydra/body
       :desc "zoom" "z" #'+hydra/text-zoom/body
       ))

(map! :map smerge-mode-map
      "s-m" #'+vc/smerge-hydra/body)

(map! :map org-mode-map
      "s-h" #'jupyter-org-hydra/body)

(+global-word-wrap-mode 1)

(global-subword-mode)

(display-time)

(setq uniquify-buffer-name-style 'post-forward-angle-brackets)

(setq confirm-kill-emacs nil)

(after! gcmh
  (setq gcmh-high-cons-threshold 33554432))

(map! :leader
      :desc "Show popup" "t p" #'+popup/toggle
      )

(load-theme 'doom-palenight t)
(setq doom-themes-enable-bold nil)
(set-face-bold-p 'bold nil)

(setq mixed-pitch-set-height t)
(setq doom-font (font-spec :family "Hack" :size 13)
      doom-variable-pitch-font (font-spec :family "Lora" :size 14)
      )
(doom/reload-font)
